DataFusion TableProvider for memory arrays

a10y opened the pull request:
Implements a readonly DataFusion `TableProvider` around an array of `DType::Struct` rows. See tests in `vortex-datafusion/lib.rs` for example of how to read from a SQL context.


* Adds new `vortex-datafusion` crate. Types are still private until we're more comfortable making them publicly exposed
* Adds new cfg-flagged converters in `vortex-dtype` to convert from `DType` to arrow `DataType` and arrow `Schema`

Support for pushing down filters against Vortex arrays will come in a FLUP tomorrow

a10y committed:
save commit

a10y committed:
save commit

a10y committed:
bump lock

a10y committed:
save

a10y committed:
more

a10y committed:
Merge remote-tracking branch 'origin/develop' into aduffy/datafusion

a10y committed:
remove broken comment

a10y committed:
fix some things, remove Cargo.lock

a10y committed:
schema

a10y committed:
small comment fixes

a10y committed:
add back Cargo lock

a10y committed:
update cargo lock

a10y commented on Cargo.toml:
```diff
@@ -41,15 +42,20 @@ arrow-data = "51.0.0"
 criterion = { version = "0.5.1", features = ["html_reports"] }
 croaring = "1.0.1"
 csv = "1.3.0"
-datafusion-common = "39.0.0"
-datafusion-expr = "39.0.0"
+datafusion = "37.1.0"
+datafusion-common = "37.1.0"
+datafusion-execution = "37.1.0"
+datafusion-expr = "37.1.0"
+datafusion-physical-expr = "37.1.0"
+datafusion-physical-plan = "37.1.0"
```
I can force everything up to `39.0.0` if i do `default-features = false`.

However, once I do that, then there are `arrow` collisions, `pyo3` collisions, and finally a whole bunch of `lance` issues.

I'm not sure if there's a way around this, or if we're stuck on this version until lance upgrades arrow and datafusion

gatesn replied:
I think the plan is to move benchmarks out, or maybe run them from Python, so things are pre-compiled

robert3005 replied:
@a10y we nuked lance for now. If we care about in the future we can figure out better setup. They don't upgrade dependencies as aggressively as we do.

gatesn commented on vortex-datafusion/src/lib.rs:
```diff
@@ -0,0 +1,323 @@
+            vortex_bail!(InvalidArgument: "only DType::Struct arrays can produce a table provider");
+        }
+
+        let arrow_schema = Schema::try_from(array.dtype())?;
```
We actually removed this conversion since it is kind of lossy. You can't tell just from a Vortex DType the exact Arrow type, e.g. is it a String or a LargeString?

What we wanted to add was `array.arrow_dtype()` possibly related to the AsArrow compute function, or add an AsArrowDType compute function or something....

a10y replied:
I think VarBin and VarBinView are the only places where this is ambiguous, and for those you can introspect the encoding details. I think it makes most sense to just implement this as a method on `Flattened`...does that sound right?

robert3005 replied:
This will happen more often - timestamp/date/time will be likely different. Similarly for list/fixedsizedlist/largelist/listview

gatesn commented on vortex-datafusion/src/lib.rs:
```diff
@@ -0,0 +1,323 @@
+
+        // Construct the RecordBatch by flattening each struct field and transmuting to an ArrayRef.
+        let struct_array = StructArray::try_from(data)
+            .map_err(|_| exec_datafusion_err!("top-level array must be Struct encoding"))?;
```
Since this is already checked, can you `.expect(...)`?

a10y replied:
Sure. I'm still a bit fuzzy where to the draw line on panicking VS returning an error for functions that are already returning Result type

robert3005 replied:
You do the check to be able to return the error and at that point if you need to unwrap the value you can panic since it should never happen (tm).

a10y replied:
done

gatesn commented on vortex-datafusion/src/lib.rs:
```diff
@@ -0,0 +1,323 @@
+        let field_order = if let Some(projection) = projection {
+            projection.clone()
+        } else {
+            (0..struct_array.names().len()).collect()
```
Does this work? Or does projection refer to the depth-first column indices? May need to hook into ArrayVisitor to count the total number of columns?

a10y replied:
From the doc comment on the `scan` method:

```
Projection
If specified, only a subset of columns should be returned, in the order specified. The projection is a set of indexes of the fields in Self::schema.
```

I checked the builtin TableProvider for Parquet files, and it defers to `SchemaRef::project(Vec<usize>)` which basically does the thing I'm doing here.

So I guess datafusion doesn't support skipping nested columns

gatesn replied:
Hmmm... that could be problematic for us!

a10y replied:
Looks like this has been outstanding for awhile: https://github.com/apache/arrow-rs/issues/5148

And yea, I suppose in the near-term, we can unnest attribute sets into top-level columns in the schema. Need to think more about that

robert3005 replied:
Oh, this is by Influx people

a10y replied:
Yea, one of the linked issues literally has a statement to the effect of "if we added this it would break IOx"

a10y replied:
i'm gonna punt this for now, it sounds like if we care about integrating spiral with DF we're going to either need to convince DF to accept PRs or we'll need to change how we represent schemas

gatesn replied:
What we could do is write a function that flattens a struct array such that the column names are complex strings, e.g. `foo.bar.[0].baz`, escaping periods.

Then we could do some fudging in Spiral if/when we need it.

gatesn commented on vortex-datafusion/src/lib.rs:
```diff
@@ -0,0 +1,323 @@
+                .map_err(|_| exec_datafusion_err!("failed to flatten struct field array"))?;
+
+            let array_ref = match field_array {
+                Flattened::Null(a) => AsArrowArray::as_arrow(&a).unwrap(),
```
`AsArrowArray::as_arrow(&field_array.into_array()).expect("Flat arrays must convert to arrow")`

Or maybe just add an `as_arrow` function to Flattened? Then can unwrap the error inside there.

a10y replied:
will add to Flattened

a10y replied:
done

gatesn commented on vortex-datafusion/src/lib.rs:
```diff
@@ -0,0 +1,323 @@
+
+        let schema = Arc::new(Schema::new(record_batch_fields));
+
+        let batch = RecordBatch::try_new(Arc::clone(&schema), record_batch)?;
```
Actually, reading all this, can you just do a single `RecordBatch::from(as_arrow(&array).map_err(...))`? 

I guess we're missing the ability to "select" columns? But we could add support to the StructArray for that.

a10y replied:
I didn't realize that there was a converter from `ArrowStructArray` -> `RecordBatch`, but also yea goal was to implement the projection pushdown here.

I'll add a `project(Vec<usize>)` method to `StructArray` and clean this up a bit

a10y replied:
done

a10y committed:
projection

a10y commented on Cargo.toml:
```diff
@@ -41,15 +42,20 @@ arrow-data = "51.0.0"
 csv = "1.3.0"
-datafusion-common = "39.0.0"
-datafusion-expr = "39.0.0"
+datafusion = "37.1.0"
+datafusion-common = "37.1.0"
+datafusion-execution = "37.1.0"
+datafusion-expr = "37.1.0"
+datafusion-physical-expr = "37.1.0"
+datafusion-physical-plan = "37.1.0"
```
will merge in develop and bump this now that lance is no longer being pulled in

a10y committed:
add simple utf8view test

a10y committed:
Merge remote-tracking branch 'origin/develop' into aduffy/datafusion

a10y committed:
more

a10y committed:
Merge remote-tracking branch 'origin/develop' into aduffy/datafusion

a10y committed:
remove test

a10y committed:
centralize flattening logic into one location

a10y committed:
things

a10y commented:
Per our huddle earlier, I've gone ahead and done the following (apologies it's split across a few commits now, if it's too hard to review I can try and stack the PRs more):

* Removed the `AsArrowArray` trait entirely, instead centralized the implementation as internal to the `Flattened` type
* Removed `VarBinView` as one of the Flattened variants. `ArrayFlatten` for `VarBinViewArra` will do a naive rebuild of the array as a fresh `VarBinArray`.
* Moved conversion functions between Vortex and Arrow schema types out of `vortex-dtype`, instead placing it as a new module within the `vortex-datafusion` crate

gatesn commented on vortex-array/src/array/struct/mod.rs:
```diff
@@ -98,6 +99,41 @@ impl StructArray {
+        let mut children = Vec::with_capacity(projection.len());
+        let mut names = Vec::with_capacity(projection.len());
+
+        let validity = self.validity().take(
```
I don't think this is right? Struct validity is actually quite rare, not sure we ever use it, but it should be preserved untouched in this operation.

a10y replied:
yep, realized this right after i pushed. fixed now.

gatesn commented on vortex-array/src/array/varbinview/mod.rs:
```diff
@@ -219,7 +220,25 @@ impl VarBinViewArray {
+        //  VarBinArray, and decoding elements from this Array one-at-a-time for insertion.
+        //  In the future, this will be removed because we will provide VarBinViewArray as a
+        //  first-class Flattened variant with zero-copy to Arrow.
+        let mut builder = VarBinBuilder::<u32>::with_capacity(self.len());
```
Is there a cast function in arrow-rs that will do this?

a10y replied:
i don't see any, may be looking in wrong place though https://github.com/apache/arrow-rs/blob/master/arrow-cast/src/cast/string.rs

robert3005 replied:
this https://github.com/apache/arrow-rs/blob/master/arrow-cast/src/cast/mod.rs#L1357?

a10y replied:
done

gatesn commented on vortex-array/src/flatten.rs:
```diff
@@ -1,14 +1,32 @@
+use std::sync::Arc;
```
Almost time to turn this into a directory. IntelliJ can right-click + promote module to directory. Then could pull out arrow logic. No strong preference

a10y replied:
i'm just gonna leave it for now. i feel like it's not quite hard enough to search through just yet...maybe in 100 lines or so

a10y commented on vortex-array/src/array/struct/mod.rs:
```diff
@@ -98,6 +99,41 @@ impl StructArray {
+        let mut children = Vec::with_capacity(projection.len());
+        let mut names = Vec::with_capacity(projection.len());
+
+        let validity = self.validity().take(
+            &PrimitiveArray::from(projection.iter().map(|idx| *idx as u64).collect::<Vec<_>>())
+                .into_array(),
+        )?;
```
this is wrong. fixing

gatesn commented on vortex-array/src/flatten.rs:
```diff
@@ -17,10 +35,189 @@ pub enum Flattened {
     Struct(StructArray),
     VarBin(VarBinArray),
-    VarBinView(VarBinViewArray),
+    // TODO(aduffy): VarBinView is being disabled until execution engines improve their
```
Maybe also mention that we only want a single canonical representation per DType. So it's not that we would enable it, it's that we would swap it out.

a10y replied:
good point, done

gatesn commented on vortex-array/src/flatten.rs:
```diff
@@ -17,10 +35,189 @@ pub enum Flattened {
+                "vortex.localdatetime" => local_date_time_to_arrow(
+                    LocalDateTimeArray::try_from(&a.into_array()).expect("localdatetime"),
+                ),
+                _ => panic!("unsupported extension dtype with ID {}", a.id().as_ref()),
```
For now. We should later be able to convert into Arrow extension array.

a10y replied:
ack

a10y committed:
fix validity, fix comment, add test

a10y committed:
Merge remote-tracking branch 'origin/develop' into aduffy/datafusion

a10y commented:
I __would__ like to rename Flatten to something else, but this PR has gotten big enough already :P

a10y committed:
cast func

a10y committed:
undo this

alamb commented:
This is so cool -- great to see this 🚀 

Let us know if there is anything we can do to help you along upstream in DataFusion

a10y commented:
Thanks Andrew, and thanks for the amazing work you're doing with the DataFusion project!

I think the two big things we're tracking are:

- Better support for StringView, which it seems like you're already tracking and the community is making progress on!
- Longer term, we think we're going to want to do projection pushdown over nested schemas, just given our data model. It seems like currently the TableProvider doesn't provide hooks for doing that.

alamb commented:
> Thanks Andrew, and thanks for the amazing work you're doing with the DataFusion project!

❤️ 

> * Better support for StringView, which it seems like you're already tracking and the community is making progress on!

Indeed -- it seem to be coming along very nicely. We are tracking in https://github.com/apache/datafusion/issues/10918

> * Longer term, we think we're going to want to do projection pushdown over nested schemas, just given our data model. It seems like currently the TableProvider doesn't provide hooks for doing that.

When you say "projection pushdown" is this what you mean: https://github.com/apache/datafusion/issues/2581 ? I just want to make sure the requests is tracked
